sql/grants.sql

```diff
@@ -0,0 +1,181 @@
+    END IF;
+
+    -- Revoke everything to begin with - looping over result sets from the revocation
+    -- cursors and running the dynamic sql; firstly schema-wide, then table-specific.
+    SET cursor_done = 0;
+    OPEN cur_revschema;
+    revschema: REPEAT
+        FETCH cur_revschema INTO vsql;
+        IF cursor_done THEN
+            LEAVE revschema;
+        END IF;
+        SET @sqlText = vsql;
+        PREPARE statement FROM @sqlText;
+        EXECUTE statement;
+        DEALLOCATE PREPARE statement;
+    UNTIL cursor_done END REPEAT revschema;
+    CLOSE cur_revschema;
+
+    SET cursor_done = 0;
+    OPEN cur_revtable;
+    revtable: REPEAT
+        FETCH cur_revtable INTO vsql;
+        IF cursor_done THEN
+            LEAVE revtable;
+        END IF;
+        SET @sqlText = vsql;
+        PREPARE statement FROM @sqlText;
+        EXECUTE statement;
+        DEALLOCATE PREPARE statement;
+    UNTIL cursor_done END REPEAT revtable;
+    CLOSE cur_revtable;
+
+    -- Set up the granular permissions as needed using the table lists in the cursors at the
+    -- top of the file
+    SET cursor_done = 0;
+    OPEN cur_select;
+    lselect: REPEAT
+        FETCH cur_select INTO tableName;
+        IF cursor_done THEN
+            LEAVE lselect;
+        END IF;
+        SET @sqlText = CONCAT('GRANT SELECT ON ', schemaname, '.', tableName, ' TO ', targetuser);
+        PREPARE statement FROM @sqlText;
+        EXECUTE statement;
+        DEALLOCATE PREPARE statement;
+    UNTIL cursor_done END REPEAT lselect;
+    CLOSE cur_select;
+
+    SET cursor_done = 0;
+    OPEN cur_update;
+    lupdate: REPEAT
+        FETCH cur_update INTO tableName;
+        IF cursor_done THEN
+            LEAVE lupdate;
+        END IF;
+        SET @sqlText = CONCAT('GRANT UPDATE ON ', schemaname, '.', tableName, ' TO ', targetuser);
+        PREPARE statement FROM @sqlText;
+        EXECUTE statement;
+        DEALLOCATE PREPARE statement;
+    UNTIL cursor_done END REPEAT lupdate;
+    CLOSE cur_update;
+
+    SET cursor_done = 0;
+    OPEN cur_insert;
+    linsert: REPEAT
+        FETCH cur_insert INTO tableName;
+        IF cursor_done THEN
+            LEAVE linsert;
+        END IF;
+        SET @sqlText = CONCAT('GRANT INSERT ON ', schemaname, '.', tableName, ' TO ', targetuser);
+        PREPARE statement FROM @sqlText;
+        EXECUTE statement;
+        DEALLOCATE PREPARE statement;
+    UNTIL cursor_done END REPEAT linsert;
+    CLOSE cur_insert;
+
+    SET cursor_done = 0;
+    OPEN cur_delete;
+    ldelete: REPEAT
+        FETCH cur_delete INTO tableName;
+        IF cursor_done THEN
+            LEAVE ldelete;
+        END IF;
+        SET @sqlText = CONCAT('GRANT DELETE ON ', schemaname, '.', tableName, ' TO ', targetuser);
+        PREPARE statement FROM @sqlText;
+        EXECUTE statement;
+        DEALLOCATE PREPARE statement;
+    UNTIL cursor_done END REPEAT ldelete;
+    CLOSE cur_delete;
+
```

Comment:
Ugh.  This feels so DRY to me, but I don't see any reasonable way to avoid having to repeat basically the same loop four times.  :|